File: CFD_Principles_and_Applications_Blazek/BookCode/Fortran/structured_1d/src/fluxroe.f
      SUBROUTINE Flux_roe( imax,ib2,a,ls,rs,f,rhs )

C *****************************************************************************
C
C  Computes convective fluxes and the dissipation according to Roe's
C  flux-difference splitting scheme and forms the residual. Source term
C  is added later.
C
C *****************************************************************************
C
C  Subroutines called: none
C
C  Functions called: Entropy_corr
C
C *****************************************************************************
C
C  Created    : Jul. 25, 1997; (c) Jiri Blazek
C  Last update: Nov. 26, 2004
C
C  This program is free software; you can redistribute it and/or
C  modify it under the terms of the GNU General Public License
C  as published by the Free Software Foundation; either version 2
C  of the License, or (at your option) any later version.
C
C  This program is distributed in the hope that it will be useful,
C  but WITHOUT ANY WARRANTY; without even the implied warranty of
C  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
C  GNU General Public License for more details.
C
C  You should have received a copy of the GNU General Public License
C  along with this program; if not, write to the Free Software
C  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
C
C *****************************************************************************

      IMPLICIT NONE
      INCLUDE 'param.inc'

C ... parameter list
      INTEGER imax, ib2
      REAL*8  a(imax), rhs(imax,3)
      REAL*8  ls(imax,3), rs(imax,3), f(imax,3)

C ... local variables
      INTEGER i
      REAL*8  fcav(3), fdiss(3)
      REAL*8  ggm1, si, rl, ul, pl, hl, qrl, rr, ur, pr, hr, qrr,
     &        rav, dd, dd1, uav, hav, q2a, c2a, cav, du, delta,
     &        h1, h2, h3, h5, eabs1, eabs2, eabs3

C ... functions
      REAL*8 Entropy_corr

C *****************************************************************************

      ggm1 = gamma/gam1

      DO i=1,ib2
        si = 0.5D0*(a(i+1)+a(i))

C ----- average of left and right convective fluxes

        rl  = ls(i,1)
        ul  = ls(i,2)
        pl  = ls(i,3)
        hl  = ggm1*pl/rl + 0.5D0*ul*ul
        qrl = ul*rl

        rr  = rs(i,1)
        ur  = rs(i,2)
        pr  = rs(i,3)
        hr  = ggm1*pr/rr + 0.5D0*ur*ur
        qrr = ur*rr

        fcav(1) = qrl    + qrr
        fcav(2) = qrl*ul + qrr*ur + pl + pr
        fcav(3) = qrl*hl + qrr*hr

C ----- dissipative fluxes

        rav   = SQRT(rl*rr)
        dd    = rav/rl
        dd1   = 1.D0/(1.D0+dd)
        uav   = (ul+dd*ur)*dd1
        hav   = (hl+dd*hr)*dd1
        q2a   = 0.5D0*uav*uav
        c2a   = gam1*(hav-q2a)
        cav   = SQRT(c2a)
        du    = ur - ul

        h1    = ABS(uav - cav)
        h2    = ABS(uav)
        h3    = ABS(uav + cav)
        delta = epsentr*cav

        eabs1 = Entropy_corr( h1,delta )
        eabs2 = Entropy_corr( h2,delta )
        eabs3 = Entropy_corr( h3,delta )

        h1 = rav*cav*du
        h2 = eabs1*(pr-pl - h1)/(2.D0*c2a)
        h3 = eabs2*(rr-rl - (pr-pl)/c2a)
        h5 = eabs3*(pr-pl + h1)/(2.D0*c2a)

        fdiss(1) = h2 + h3 + h5
        fdiss(2) = h2*(uav-cav) + h3*uav + h5*(uav+cav)
        fdiss(3) = h2*(hav-cav*uav) + h3*q2a + h5*(hav+cav*uav)

C ----- total fluxes at i+1/2

        f(i,1) = 0.5D0*(fcav(1)-fdiss(1))*si
        f(i,2) = 0.5D0*(fcav(2)-fdiss(2))*si
        f(i,3) = 0.5D0*(fcav(3)-fdiss(3))*si
      ENDDO

C --- sum of fluxes = RHS -----------------------------------------------------

      DO i=2,ib2
        rhs(i,1) = f(i,1) - f(i-1,1)
        rhs(i,2) = f(i,2) - f(i-1,2)
        rhs(i,3) = f(i,3) - f(i-1,3)
      ENDDO

      RETURN
      END

C =============================================================================

      REAL*8 FUNCTION Entropy_corr( z,d )
C
C  Entropy correction function
C
      IMPLICIT NONE

      REAL*8 z, d

      IF (z .GT. d) THEN
        Entropy_corr = z
      ELSE
        Entropy_corr = 0.5D0*(z*z+d*d)/d
      ENDIF

      RETURN
      END

File: Aircraft_Aerodynamic_Design_with_Computational_Software_Rizzi/ExercisesLabs/Ch4/Ch4lib/flux_roe.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                   %
%                         DEMOFLOW                                  %
%                                                                   %
%       CALCULATION OF QUASI-1D INVISCID FLOW IN A TUBE/NOZZLE      %
%       ======================================================      %
%             Alexander von Essen, Created Apr. 5 2005              %
%                   Last modified: May 10 2005                      %
%                                                                   %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%                                                                   %
% Name: flux_roe.m                                                  %
% Purpose: Calculates convective flux and dissipation using         %
%          Roe's flux-difference splitting scheme.                  %
%          Source term added later.                                 %
% Called by: solveexplicit.m                                        %
% Calls: entropy_correction.m                                       %
%                                                                   %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [handles] = flux_roe(handles)
ib2 = handles.Data.Nodes - 1;
a = handles.Data.a;
p = handles.Data.p;
gamma = handles.Data.gamma;
ls = handles.Data.ls;
rs = handles.Data.rs;
epsentr = handles.Data.epsentr;

nnodes=ib2+1;

fdiss = zeros(nnodes,3);
fcav  = zeros(nnodes,3);

si = 0.5.*(a(2:end) + a(1:ib2));

rl  = ls(1:ib2,1);
ul  = ls(1:ib2,2);
pl  = ls(1:ib2,3);
hl  = (gamma/(gamma-1)).*(pl./rl) + 0.5.*ul.^2;
qrl = ul.*rl;

rr  = rs(1:ib2,1);
ur  = rs(1:ib2,2);
pr  = rs(1:ib2,3);
hr  = (gamma/(gamma-1)).*(pr./rr) + 0.5.*ur.^2;
qrr = ur.*rr;

fcav(1:ib2,1) = qrl + qrr;
fcav(1:ib2,2) = qrl.*ul + qrr.*ur + pl + pr;
fcav(1:ib2,3) = qrl.*hl + qrr.*hr;

rav = sqrt(rl.*rr);
dd  = rav./rl;
dd1 = 1./(1+dd);

uav = (ul+dd.*ur).*dd1;
hav = (hl+dd.*hr).*dd1;
q2a = 0.5.*uav.^2;
c2a = (gamma-1).*(hav-q2a);
cav = sqrt(c2a);
du_b  = ur - ul;

% Eigenvalues of the Roe matrix.
h1  = abs(uav - cav);
h2  = abs(uav);
h3  = abs(uav + cav);

%%% New block %%%
% newbl = 0;
% if newbl
%     h1 = [];
%     h2 = [];
%     h3 = [];
%     
%     %save AA AA
%     
%     for n = 1:length(AA)
%         D = eig(AA(:,:,n));
%         D = sort(abs(D));
%         %   D = abs(D);
%         h1 = [h1; D(1)];
%         h2 = [h2; D(2)];
%         h3 = [h3; D(3)];
%     end
% end

delta = epsentr.*h3;

eabs1 = entropy_correction(h1,delta);
eabs2 = entropy_correction(h2,delta);
eabs3 = entropy_correction(h3,delta);

h1 = rav.*cav.*du_b;
% JO 1905 -
h2 = eabs1.*(pr-pl-h1)./(2.*c2a);
h3 = eabs2.*(rr-rl-(pr-pl)./c2a);
h5 = eabs3.*(pr-pl+h1)./(2.*c2a);
jos = 1;
fdiss(1:ib2,1) = h2 + h3 + jos*h5;
fdiss(1:ib2,2) = h2.*(uav-cav) + h3.*uav + jos*h5.*(uav+cav);
%fdiss(1:ib2,3) = h2.*(hav-cav.*uav) + h3.*q2a + h5.*(hav+cav+uav);
% JO 1905                                               ----V----
fdiss(1:ib2,3) = h2.*(hav-cav.*uav) + h3.*q2a + jos*h5.*(hav+cav.*uav);


du(1:ib2,1) = 0.5.*(fcav(1:ib2,1)-fdiss(1:ib2,1)).*si;
du(1:ib2,2) = 0.5.*(fcav(1:ib2,2)-fdiss(1:ib2,2)).*si;
du(1:ib2,3) = 0.5.*(fcav(1:ib2,3)-fdiss(1:ib2,3)).*si;

rhs=zeros(nnodes,3);

rhs(2:ib2,1) = diff(du(1:ib2,1));
rhs(2:ib2,2) = diff(du(1:ib2,2));
rhs(2:ib2,3) = diff(du(1:ib2,3));

handles.Data.rhs = rhs;
